packages/graphql-codegen-supermassive-typed-document-node-plugin/src/visitor.ts

```diff
@@ -80,23 +75,89 @@ export class TypeScriptDocumentNodesVisitor extends ClientSideBaseVisitor<
       ) as string;
       this._imports.add(tagImport);
     }
+
+    if (this.config.astType === "Lazy") {
+      const { runtimeParseImport } = rawConfig;
+      if (!validateRuntimeParseImport(runtimeParseImport)) {
+        throw new Error(
+          `graphql-codegen-supermassive-typed-document-node-plugin: astType="Lazy" requires valid runtimeParseImport in the config`,
+        );
+      }
+
+      const tagImport = this._generateImport(
+        runtimeParseImport,
+        "runtimeParse",
+        false,
+      ) as string;
+      this._imports.add(tagImport);
+    }
+  }
+
+  protected getDocumentNodeSignature(
+    resultType: string,
+    variablesTypes: string,
+    node: FragmentDefinitionNode | OperationDefinitionNode,
+  ) {
+    if (
+      this.config.documentMode === DocumentMode.documentNode ||
+      this.config.documentMode === DocumentMode.documentNodeImportFragments ||
+      this.config.documentMode === DocumentMode.graphQLTag
+    ) {
+      return ` as unknown as DocumentNode<${resultType}, ${variablesTypes}>`;
+    }
+
+    return super.getDocumentNodeSignature(resultType, variablesTypes, node);
   }
 
   protected _gql(
     node: FragmentDefinitionNode | OperationDefinitionNode,
   ): string {
-    if (this.config.supermassiveDocumentNodeConditional) {
-      const supermassive = this._render(node, true);
-      const standard = this._render(node, false);
-      return `(${this.config.supermassiveDocumentNodeConditional}\n? ${supermassive}\n: ${standard})`;
+    switch (this.config.astType) {
+      case "Lazy":
+        return this._renderLazy(node);
+
+      case "SupermassiveV3":
+        return this._render(node, true);
+
+      case "Standard":
+      default:
+        return this._render(node, false);
     }
-    return this._render(
-      node,
-      this.config.supermassiveDocumentNodeOutputType !== "NONE",
-    );
   }
 
-  protected _render(
+  private _renderLazy(
+    node: FragmentDefinitionNode | OperationDefinitionNode,
+  ): string {
+    const fragments = this._transformFragments(node);
+    const doc = this._prepareDocument(`
+    ${
+      print(node)
+        .split("\\")
+        .join("\\\\") /* Re-escape escaped values in GraphQL syntax */
+    }
+    ${this._includeFragments(fragments)}
+    `);
+
+    const gqlObj = parse(doc, { noLocation: true });
+    return `
+{
+  kind: "Document",
+  get definitions() {
+    const document = runtimeParse('${print(gqlObj).replace(/[\n\r ]+/g, " ")}');
```

Comment:
Oh, also this `.replace(/[\n\r ]+/g)` is potentially breaking (e.g. for inline string arguments with spaces). We need a custom version of `print` with a different set of whitespace characters instead. Sadly, graphql-js doesn't expose any options for print, so we will have to fork the printer:

https://github.com/graphql/graphql-js/blob/16.x.x/src/language/printer.ts

Reply:
added custom printer to resolve this